File: 21558 - Nithin/week 3/918-maximum-sum-circular-subarray/918-maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int n = nums.size();
        if(n == 1)
            return nums[0];
        
        int arr_sum =0;
        for(int i = 0;i < n;i++)
            arr_sum += nums[i];
        
        int curr_max = nums[0], max_so_far = nums[0];
        int curr_min = nums[0], min_so_far = nums[0];
        
        for(int j = 1;j < n;j++){
            curr_max = max(curr_max + nums[j], nums[j]);
            max_so_far = max(curr_max, max_so_far);
            
            curr_min = min(curr_min + nums[j], nums[j]);
            min_so_far = min(curr_min, min_so_far);
        }
        
        if(min_so_far == arr_sum)
            return max_so_far;
        
        return max(max_so_far, arr_sum - min_so_far );
        
    }
};

File: 21560 - Sai Suraj/week 3/918_Maximum Sum Circular Subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        msf = float('-inf')
        s = 0
        
        for i in nums:
            s+=i
            if s>msf:
                msf = s
            if s<0:
                s = 0
        
        s = 0
        minsf = float('inf')
        for i in nums[1:-1]:
            s+=i
            if s<minsf:
                minsf = s
            if s>0:
                s = 0
                
        return max(msf,sum(nums)-minsf)
